daemon/restapi/config.go

```diff
@@ -0,0 +1,427 @@
+
+type ConfigModifyEventHandler struct {
+	logger logrus.FieldLogger
+
+	datapathRegenTrigger *trigger.Trigger
+	// event queue for serializing configuration updates to the daemon.
+	configModifyQueue *eventqueue.EventQueue
+
+	datapath        datapath.Datapath
+	policy          *policy.Repository
+	endpointManager endpointmanager.EndpointManager
+	l7Proxy         *proxy.Proxy
+}
+
+func (h *ConfigModifyEventHandler) datapathRegen(reasons []string) {
+	reason := strings.Join(reasons, ", ")
+
+	regenerationMetadata := &regeneration.ExternalRegenerationMetadata{
+		Reason:            reason,
+		RegenerationLevel: regeneration.RegenerateWithDatapath,
+	}
+	h.endpointManager.RegenerateAllEndpoints(regenerationMetadata)
+}
+
+func (h *ConfigModifyEventHandler) configModify(params daemonapi.PatchConfigParams, resChan chan interface{}) {
+	cfgSpec := params.Configuration
+
+	om, err := option.Config.Opts.Library.ValidateConfigurationMap(cfgSpec.Options)
+	if err != nil {
+		msg := fmt.Errorf("invalid configuration option: %w", err)
+		resChan <- api.Error(daemonapi.PatchConfigBadRequestCode, msg)
+		return
+	}
+
+	// Serialize configuration updates to the daemon.
+	option.Config.ConfigPatchMutex.Lock()
+
+	// Track changes to daemon's configuration
+	var changes int
+	var policyEnforcementChanged bool
+	// Copy old configurations for potential reversion
+	oldEnforcementValue := policy.GetPolicyEnabled()
+	oldConfigOpts := option.Config.Opts.DeepCopy()
+	oldEpConfigOpts := make(option.OptionMap, len(om))
+	for k := range om {
+		oldEpConfigOpts[k] = oldConfigOpts.Opts[k]
+	}
+
+	// Only update if value provided for PolicyEnforcement.
+	if enforcement := cfgSpec.PolicyEnforcement; enforcement != "" {
+		switch enforcement {
+		case option.NeverEnforce, option.DefaultEnforcement, option.AlwaysEnforce:
+			// If the policy enforcement configuration has indeed changed, we have
+			// to regenerate endpoints and update daemon's configuration.
+			if enforcement != oldEnforcementValue {
+				h.logger.Debug("configuration request to change PolicyEnforcement for daemon")
+				changes++
+				policy.SetPolicyEnabled(enforcement)
+				policyEnforcementChanged = true
+			}
+
+		default:
+			msg := fmt.Errorf("invalid option for PolicyEnforcement %s", enforcement)
+			h.logger.Warn(msg)
+			option.Config.ConfigPatchMutex.Unlock()
+			resChan <- api.Error(daemonapi.PatchConfigBadRequestCode, msg)
+			return
+		}
+		h.logger.Debug("finished configuring PolicyEnforcement for daemon")
+	}
+
+	changes += option.Config.Opts.ApplyValidated(om, h.changedOption, nil)
+	h.endpointManager.OverrideEndpointOpts(om)
+
+	h.logger.WithField("count", changes).Debug("Applied changes to daemon's configuration")
+	option.Config.ConfigPatchMutex.Unlock()
+
+	if changes > 0 {
+		// Only recompile if configuration has changed.
+		h.logger.Debug("daemon configuration has changed; recompiling base programs")
+		if err := h.datapath.Orchestrator().Reinitialize(context.TODO()); err != nil {
```

Comment:
TODO: replace `context.TODO()` with a context that spans the lifecycle of the cell (~ agent). Previously this was using `Daemon.ctx`.

Reply:
fixed. the event handler now also contains a context as field (analogous to the daemon). Even though not the nicest thing, i'd like to keep the logic the same in this PR.